src/wms/src/test/java/org/geoserver/wms/capabilities/GetCapabilitiesTransformerTest.java

```diff
@@ -415,6 +431,89 @@ private void checkVendorSpecificCapsProviders(GetCapabilitiesTransformer tr) thr
         assertEquals(1, list.getLength());
     }
 
+    /**
+     * Gets a capabilities document as an XML string Also, adds: <!DOCTYPE WMT_MS_Capabilities
+     * SYSTEM "WMS_MS_Capabilities.dtd"> at the top of the document.
+     */
+    String getCapabilitiesXML() throws Exception {
+        // WMS wms = new WMS(getGeoServer());
+        WMS wms = (WMS) applicationContext.getBean("wms");
+
+        GetCapabilitiesTransformer tr =
+                new GetCapabilitiesTransformer(wms, baseUrl, mapFormats, legendFormats, null);
+
+        Document dom = WMSTestSupport.transform(req, tr);
+        TransformerFactory ttf = TransformerFactory.newInstance();
+        Transformer trans = ttf.newTransformer();
+        StringWriter sw = new StringWriter();
+        trans.transform(new DOMSource(dom), new StreamResult(sw));
+
+        String xml = sw.toString();
+        xml =
+                xml.replace(
+                        "<?xml version=\"1.0\" encoding=\"UTF-8\"?>",
+                        "<?xml version=\"1.0\" encoding=\"UTF-8\"?>"
+                                + "<!DOCTYPE WMT_MS_Capabilities SYSTEM \"WMS_MS_Capabilities.dtd\">");
+
+        return xml;
+    }
+
+    @Test
+    public void testValidatesAgainstDTD() throws Exception {
+        // get a capabilities document
+        String getCapXML = getCapabilitiesXML();
+
+        // get the wms 1.1.1 DTD
+        URL dtdURL =
+                GetCapabilitiesTransformer.class.getResource(
+                        "/schemas/wms/1.1.1/wms_ms_capabilities.dtd");
+        String dtd = Resources.toString(dtdURL, StandardCharsets.UTF_8);
+
+        try (InputStream dtdInputStream = new ByteArrayInputStream(dtd.getBytes())) {
+
+            // parse and validate the capabilities document against the DTD
+            DocumentBuilderFactory factory = DocumentBuilderFactory.newInstance();
+            factory.setValidating(true);
+            factory.setNamespaceAware(true);
+
+            DocumentBuilder builder = factory.newDocumentBuilder();
+
+            // Normally, the DTD would downloaded from the internet.  We don't want to do that, so
+            // we tell the parse to use our DTD instead of downloading it.
+            builder.setEntityResolver(
+                    new EntityResolver() {
+                        @Override
+                        public InputSource resolveEntity(String publicId, String systemId)
+                                throws SAXException, IOException {
+                            if (systemId.endsWith("WMS_MS_Capabilities.dtd")) {
+                                return new InputSource(dtdInputStream);
+                            }
+                            return null;
+                        }
+                    });
```

Comment:
This looks legit; and I would expect both factory and builder to be isolated. 

Cannot tell why other threads are unhappy and failing.